theano/gpuarray/opt.py

```diff
@@ -215,6 +254,272 @@ def apply(self, fgraph):
+                            node.outputs)
+                    finally:
+                        t_opt2 = time.time()
+                    if new_ops:
+                        break
+            if not new_ops:
+                newnode = node.clone_with_new_inputs([mapping.get(i)
+                                                      for i in node.inputs])
+                outputs = newnode.outputs
+            elif isinstance(new_ops, (tuple, list)):
+                outputs = []
+                for o in new_ops:
+                    if o.owner and isinstance(o.owner.op, HostFromGpu):
+                        outputs.append(o.owner.inputs[0])
+                    else:
+                        outputs.append(o)
+            elif isinstance(new_ops, theano.Variable):
+                if new_ops.owner and isinstance(new_ops.owner.op, HostFromGpu):
+                    outputs = new_ops.owner.inputs
+                else:
+                    outputs = [new_ops]
+            else:
+                outputs = new_ops(*[mapping[i] for i in node.inputs],
+                                  return_list=True)
+
+            if new_ops:
+                node_created[lopt] += len(theano.gof.graph.ops([mapping[i] for i in node.inputs], outputs))
+                self.new_opts.append(lopt)
+                time_opts[lopt] = t_opt2 - t_opt
+
+            for new_o, old_o in zip(outputs, node.outputs):
+                mapping[old_o] = new_o
+
+        new_nodes = []
+        for o in fgraph.outputs:
+            new_o = mapping[o]
+            if new_o.type != o.type:
+                assert isinstance(o.type, tensor.TensorType)
+                assert isinstance(new_o.type, GpuArrayType)
+                new_o = host_from_gpu(new_o)
+            new_nodes.append(new_o)
+        fgraph.replace_all_validate(zip(fgraph.outputs, new_nodes))
+
+        end_nb_nodes = len(fgraph.apply_nodes)
+
+        return (self, start_nb_nodes, end_nb_nodes, max_nb_nodes, io_toposort_timing,
+                nb_nodes, time_opts, node_created)
+
+    @staticmethod
+    def print_profile(stream, prof, level=0):
+        (opt, start_nb_nodes, end_nb_nodes, max_nb_nodes, io_toposort_timing,
+         nb_nodes, time_opts, node_created) = prof
+        blanc = ('    ' * level)
+        print(blanc, "GraphToGPUOptimizer", end=' ', file=stream)
+        print(blanc, getattr(opt, "name",
+                             getattr(opt, "__name__", "")), file=stream)
+        print(blanc, "  nb nodes (start, end,  max) %d %d %d" % (
+            start_nb_nodes, end_nb_nodes, max_nb_nodes), file=stream)
+        print(blanc, "  time io_toposort %.3fs" % sum(
+            io_toposort_timing), file=stream)
+
+        s = sum([time_opts[o] for o in opt.new_opts])
+        
+        print(blanc, "  time in local optimizers %.3fs" % s, file=stream)
+
+        # Build a dictionary of opt and time taken
+        opt_time_dict = dict()
+        for o in opt.new_opts:
```

Comment:
The problem is here. time_opts is already a dict like this : opt -> total time for this opt.

new_opts contain the number of time each opt was applied. Just delete new_opts and use time_opts as opt_time_dict.